src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Decoder.cs

```diff
@@ -558,40 +591,44 @@ private static OperationStatus DecodeWithWhiteSpaceBlockwise(ReadOnlySpan<byte>
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static int GetPaddingCount(ref byte ptrToLastElement)
+        private static int GetPaddingCount<TBase64Decoder>(ref byte ptrToLastElement)
+            where TBase64Decoder : IBase64Decoder<byte>
         {
             int padding = 0;
 
-            if (ptrToLastElement == EncodingPad) padding++;
-            if (Unsafe.Subtract(ref ptrToLastElement, 1) == EncodingPad) padding++;
+            if (TBase64Decoder.IsValidPadding(ptrToLastElement))
+            {
+                padding++;
+            }
+
+            if (TBase64Decoder.IsValidPadding(Unsafe.Subtract(ref ptrToLastElement, 1)))
+            {
+                padding++;
+            }
 
             return padding;
         }
 
-        private static OperationStatus DecodeWithWhiteSpaceFromUtf8InPlace(Span<byte> utf8, ref int destIndex, uint sourceIndex)
+        private static OperationStatus DecodeWithWhiteSpaceFromUtf8InPlace<TBase64Decoder>(Span<byte> source, ref int destIndex, uint sourceIndex)
+            where TBase64Decoder : IBase64Decoder<byte>
         {
-            const int BlockSize = 4;
+            int BlockSize = Math.Min(source.Length - (int)sourceIndex, 4);
```

Comment:
For `Base64Url` when the source is not multiple of 4 we need to adjust the buffer size accordingly, the `BlockSize` value used to fill the whitespace with padding (row 653) so that it could decoded correctly (in case remaining bytes were decodable or valid)